metamorph/callbacker.go

```diff
@@ -0,0 +1,94 @@
+package metamorph
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log/slog"
+	"net/http"
+	"time"
+
+	"github.com/bitcoin-sv/arc/api"
+	"github.com/bitcoin-sv/arc/metamorph/metamorph_api"
+	"github.com/bitcoin-sv/arc/metamorph/store"
+	"github.com/ordishs/go-utils"
+)
+
+const (
+	CallbackTries    = 5
+	CallbackInterval = 30
+)
+
+func SendCallback(logger *slog.Logger, s store.MetamorphStore, tx *store.StoreData) {
+	sleepDuration := CallbackInterval
+	for i := 1; i < CallbackTries; i++ {
+		statusString := metamorph_api.Status(tx.Status).String()
+		blockHash := ""
+		if tx.BlockHash != nil {
+			blockHash = utils.ReverseAndHexEncodeSlice(tx.BlockHash.CloneBytes())
+		}
+
+		logger.Info("sending callback for transaction", slog.String("token", tx.CallbackToken), slog.String("hash", tx.Hash.String()), slog.String("url", tx.CallbackUrl), slog.Uint64("block height", tx.BlockHeight), slog.String("block hash", blockHash))
+
+		status := &api.TransactionStatus{
+			BlockHash:   &blockHash,
+			BlockHeight: &tx.BlockHeight,
+			TxStatus:    &statusString,
+			Txid:        tx.Hash.String(),
+			Timestamp:   time.Now(),
+		}
+		statusBytes, err := json.Marshal(status)
+		if err != nil {
+			logger.Error("Couldn't marshal status  - ", err)
+			return
+		}
+		statusBuffer := bytes.NewBuffer(statusBytes)
+
+		var request *http.Request
+		request, err = http.NewRequest("POST", tx.CallbackUrl, statusBuffer)
+		if err != nil {
+			logger.Error("Couldn't marshal status  - ", errors.Join(err, fmt.Errorf("failed to post callback for transaction id %s", tx.Hash)))
+			return
+		}
+		request.Header.Set("Content-Type", "application/json; charset=UTF-8")
+		if tx.CallbackToken != "" {
+			request.Header.Set("Authorization", "Bearer "+tx.CallbackToken)
+		}
+
+		// default http client
+		httpClient := http.Client{}
```

Comment:
httpClient := &http.Client {
Timeout: 5 *time.Second
}
seems a bit more idiomatic to me 😊